### PR TITLE
util, bin: push prometheus metrics

### PR DESCRIPTION
**Pull request by overvenus**
How to:

 1. start a prometheus [pushgateway]
 2. start a prometheus server and pull metrics from [pushgateway]
 3. start a tikv and let `metric.address` be the address of [pushgateway]

[pushgateway]: https://github.com/prometheus/pushgateway

### CODE REVIEW

File: etc/config-template.toml [hunk 1]
```diff
@@ -30,8 +30,12 @@ end-point-concurrency = 8
 # capacity = 0 # 0 is unlimited.
 
 [metric]
-# prometheus client report interval, set 0s to disable prometheus.
-report-interval = "5m"
+# prometheus client push interval, set 0s to disable prometheus.
+push-interval = "15s"
+# prometheus pushgateway address.
+address = "127.0.0.1:9091"
```
Comment (siddontang) [inline etc/config-template.toml]: Default should be empty. Add "127.0.0.1:9091" in above comment.

**Comment by siddontang** [comment]
@overvenus CI failed
Seem you should update travis rust version.

File: etc/config-template.toml [hunk 2]
```diff
@@ -30,8 +30,12 @@ end-point-concurrency = 8
 # capacity = 0 # 0 is unlimited.
 
 [metric]
-# prometheus client report interval, set 0s to disable prometheus.
-report-interval = "5m"
+# prometheus client push interval, set 0s to disable prometheus.
+interval = "15s"
+# prometheus pushgateway address, leaves it empty will disable prometheus.
+address = ""
+# prometheus client push job name.
```
Comment (siddontang) [inline etc/config-template.toml]: should tell user that the real job name will contain node id.

**Comment by siddontang** [comment]
Rest LGTM

PTAL @BusyJay @disksing

File: travis-build/Makefile [hunk 3]
```diff
@@ -53,7 +53,7 @@ ${LOCAL_DIR}/bin/pd-server:
 	cp bin/pd-server ${LOCAL_DIR}/bin/pd-server
 
 prepare_rust:
-	sh ~/rust/lib/rustlib/uninstall.sh && sh ~/rust-installer/rustup.sh --date=2016-09-05 --prefix=~/rust --disable-sudo --channel=nightly
+	sh ~/rust/lib/rustlib/uninstall.sh && sh ~/rust-installer/rustup.sh --date=2016-09-10 --prefix=~/rust --disable-sudo --channel=nightly
```
Comment (BusyJay) [inline travis-build/Makefile]: Why upgrade `rustc`?

Comment (overvenus) [inline travis-build/Makefile]: I upgraded clippy to `0.0.90`, `rustc 09-05` can't compile it. Are there any problems of upgrading？Or should I roll clippy back?

Comment (BusyJay) [inline travis-build/Makefile]: It's ok. But normally we create a separate pr to upgrade the `rustc` and clippy.

File: etc/config-template.toml [hunk 4]
```diff
@@ -30,8 +30,12 @@ end-point-concurrency = 8
 # capacity = 0 # 0 is unlimited.
 
 [metric]
-# prometheus client report interval, set 0s to disable prometheus.
-report-interval = "5m"
+# prometheus client push interval, set 0s to disable prometheus.
+interval = "15s"
+# prometheus pushgateway address, leaves it empty will disable prometheus.
```
Comment (BusyJay) [inline etc/config-template.toml]: Do we need to fallback to log instead?

Comment (siddontang) [inline etc/config-template.toml]: no need now, we can send SIGUSR1.

Comment (BusyJay) [inline etc/config-template.toml]: Then I guess here is not to disable prometheus but disable pushing.

Comment (siddontang) [inline etc/config-template.toml]: yes, a wrong comment.

Comment (BusyJay) [inline etc/config-template.toml]: By the way, if I use SIGUSR1 to dump metrics, will it still be pushed to remote? Or the dumped metrics will be excluded to the remote?

**Comment by siddontang** [comment]
Ping @overvenus

File: etc/config-template.toml (hunk 4 above)
Comment (overvenus) [inline etc/config-template.toml]: Yes, they will be pushed. The dumped metrics are identical to the pushd.

File: etc/config-template.toml [hunk 5]
```diff
@@ -30,9 +30,9 @@ end-point-concurrency = 8
 # capacity = 0 # 0 is unlimited.
 
 [metric]
-# prometheus client push interval, set 0s to disable prometheus.
+# prometheus client push interval, set 0s to disable prometheus from pushing.
 interval = "15s"
-# prometheus pushgateway address, leaves it empty will disable prometheus.
+# prometheus pushgateway address, leaves it empty will disable prometheus from pushing.
```
Comment (siddontang) [inline etc/config-template.toml]: @queenypingcap please review the comment.

File: etc/config-template.toml [hunk 6]
```diff
@@ -30,8 +30,12 @@ end-point-concurrency = 8
 # capacity = 0 # 0 is unlimited.
 
 [metric]
-# prometheus client report interval, set 0s to disable prometheus.
-report-interval = "5m"
+# prometheus client push interval, set 0s to disable prometheus from pushing.
```
Comment (queenypingcap) [inline etc/config-template.toml]: +# the Prometheus client push interval. Set the value to 0s to stop Prometheus from pushing.

File: etc/config-template.toml [hunk 7]
```diff
@@ -30,8 +30,12 @@ end-point-concurrency = 8
...
+interval = "15s"
+# prometheus pushgateway address, leaves it empty will disable prometheus from pushing.
```
Comment (queenypingcap) [inline etc/config-template.toml]: +# the Prometheus pushgateway address. Leaving it empty stops Prometheus from pushing.

File: etc/config-template.toml [hunk 8]
```diff
@@ -30,8 +30,12 @@ end-point-concurrency = 8
...
+address = ""
+# prometheus client push job name. Notice: A node id will automatically appends, e.g., "tikv_1".
```
Comment (queenypingcap) [inline etc/config-template.toml]: +# the Prometheus client push job name. Note: A node id will automatically append, e.g., "tikv_1".

File: etc/config-template.toml [hunk 9]
```diff
@@ -30,11 +30,11 @@ end-point-concurrency = 8
 # capacity = 0 # 0 is unlimited.
 
 [metric]
-# prometheus client push interval, set 0s to disable prometheus from pushing.
+# the Prometheus client push interval. Setting the value to 0s stops Prometheus from pushing.
```
Comment (siddontang) [inline etc/config-template.toml]: stops Prometheus client from pushing.

**Comment by siddontang** [comment]
LGTM
PTAL @BusyJay

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**